packages/core/src/engine/BBox.ts

```diff
@@ -0,0 +1,142 @@
+import { Pt2, VarAD } from "types/ad";
+import { add, constOf, div, neg, ops } from "./Autodiff";
+
+export class BBox {
+  public readonly w: VarAD;
```

Comment:
Thanks for factoring this out, it is much cleaner! 

If the OO style used just for the purpose of lazily evaluating corners of the bbox? If so, I actually have to backtrack here... on second thought, unnecessary computation is not going to show up in the energy/gradient graph if it is not ultimately used in the energy. (Unnecessary computation is still a problem if it *is* ultimately used in the energy.) 

So this is not a problem, and I think a non-OO style might be cleaner as we don't use it elsewhere in the codebase (rather, we use typedefs paired with functions).

@wodeni curious about your thoughts on introducing OO style into the codebase as here?

Reply:
Yeah the purpose of OO here was just to have the lazy computation. It can be easily refactored to just a namespace if laziness isn't an issue here.

Reply:
Yeah, if you don't mind, I'd prefer a namespace as it's cleaner

Reply:
This is a very minor point, but as I'm doing the refactor I'm reminded that the other benefit of using a class is that you can write nontrivial getters that look like properties. In the namespace version you have to choose either (i) to pre-compute `corners`, which gives BBox a public `corners` property but confusingly things like `edges` are still functions or (ii) computing `corners` every time, but having a consistent interface for all but the basic three properties of `w`, `h`, and `center`. I'm going to do the second one, because I'm assuming the computation graph execution time dwarfs the execution time of re-constructing corners.

Reply:
refactored!